File: MainProject/Scenes/HexPlanet/Struct/HexTileDataOverrider.cs
using System.Collections.Generic;
using ZeromaXsPlaygroundProject.Scenes.Framework.Dependency;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Entity;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Enum;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Service;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Util;

namespace ZeromaXsPlaygroundProject.Scenes.HexPlanet.Struct;

public enum OptionalToggle
{
    Ignore,
    Yes,
    No
}

public struct HexTileDataOverrider
{
    public HexTileDataOverrider() => InitServices();

    #region 服务

    private static IPlanetSettingService _planetSettingService;

    private static void InitServices()
    {
        _planetSettingService ??= Context.GetBean<IPlanetSettingService>();
    }

    #endregion

    public bool EditMode;
    public bool ApplyTerrain;
    public int ActiveTerrain;
    public bool ApplyElevation;
    public int ActiveElevation;
    public bool ApplyWaterLevel;
    public int ActiveWaterLevel;
    public int BrushSize;
    public OptionalToggle RiverMode;
    public OptionalToggle RoadMode;
    public bool ApplyUrbanLevel;
    public int ActiveUrbanLevel;
    public bool ApplyFarmLevel;
    public int ActiveFarmLevel;
    public bool ApplyPlantLevel;
    public int ActivePlantLevel;
    public OptionalToggle WalledMode;
    public bool ApplySpecialIndex;
    public int ActiveSpecialIndex;
    public HashSet<Tile> OverrideTiles = [];

    public bool IsOverrideTile(Tile tile) => EditMode && OverrideTiles.Contains(tile);
    public bool IsOverrideNoRiver(Tile tile) => IsOverrideTile(tile) && RiverMode == OptionalToggle.No;
    public bool IsOverrideNoRoad(Tile tile) => IsOverrideTile(tile) && RoadMode == OptionalToggle.No;

    public int Elevation(Tile tile) =>
        // 现在低于陆地高度的不再绘制低于高度的部分
        IsOverrideTile(tile) && ApplyElevation && ActiveElevation > tile.Data.Elevation
            ? ActiveElevation
            : tile.Data.Elevation;

    public int WaterLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyWaterLevel ? ActiveWaterLevel : tile.Data.WaterLevel;

    public bool IsUnderwater(Tile tile) =>
        IsOverrideTile(tile) ? WaterLevel(tile) > Elevation(tile) : tile.Data.IsUnderwater;

    public float StreamBedY(Tile tile) =>
        IsOverrideTile(tile)
            ? (Elevation(tile) + HexMetrics.StreamBedElevationOffset) * _planetSettingService.UnitHeight
            : tile.Data.StreamBedY;

    public float RiverSurfaceY(Tile tile) =>
        IsOverrideTile(tile)
            ? (Elevation(tile) + HexMetrics.WaterElevationOffset) * _planetSettingService.UnitHeight
            : tile.Data.RiverSurfaceY;

    public float WaterSurfaceY(Tile tile) =>
        IsOverrideTile(tile)
            ? (WaterLevel(tile) + HexMetrics.WaterElevationOffset) * _planetSettingService.UnitHeight
            : tile.Data.WaterSurfaceY;

    public bool HasRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasRiver;
    public bool HasIncomingRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasIncomingRiver;
    public bool HasOutgoingRiver(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasOutgoingRiver;
    public bool HasRiverBeginOrEnd(Tile tile) => !IsOverrideNoRiver(tile) && tile.Data.HasRiverBeginOrEnd;

    public bool HasRiverThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRiver(tile) && tile.Data.HasRiverThroughEdge(idx);

    public bool HasIncomingRiverThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRiver(tile) && tile.Data.HasIncomingRiverThroughEdge(idx);

    public bool HasRoads(Tile tile) => !IsOverrideNoRoad(tile) && tile.Data.HasRoads;

    public bool HasRoadThroughEdge(Tile tile, int idx) =>
        !IsOverrideNoRoad(tile) && tile.Data.HasRoadThroughEdge(idx);

    public bool Walled(Tile tile) =>
        IsOverrideTile(tile) && WalledMode != OptionalToggle.Ignore
            ? WalledMode == OptionalToggle.Yes
            : tile.Data.Walled;

    public HexEdgeType GetEdgeType(Tile tile1, Tile tile2) =>
        HexMetrics.GetEdgeType(Elevation(tile1), Elevation(tile2));

    public int UrbanLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyUrbanLevel ? ActiveUrbanLevel : tile.Data.UrbanLevel;

    public int FarmLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyFarmLevel ? ActiveFarmLevel : tile.Data.FarmLevel;

    public int PlantLevel(Tile tile) =>
        IsOverrideTile(tile) && ApplyPlantLevel ? ActivePlantLevel : tile.Data.PlantLevel;

    public int SpecialIndex(Tile tile) =>
        IsOverrideTile(tile) && ApplySpecialIndex ? ActiveSpecialIndex : tile.Data.SpecialIndex;

    public bool IsSpecial(Tile tile) => SpecialIndex(tile) > 0;
}

File: TerraObserver/TO.Domains/Functions/PlanetHuds/HexTileDataOverrider.fs
namespace TO.Domains.Functions.PlanetHuds

open TO.Domains.Functions.HexMetrics
open TO.Domains.Functions.HexSpheres.Components.Tiles
open TO.Domains.Types.HexSpheres.Components.Tiles
open TO.Domains.Types.PlanetHuds

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-29 21:26:29
module HexTileDataOverrider =
    let isOverridingTileConnection (tile: int) (neighbor: int) (this: HexTileDataOverrider) =
        this.EditMode
        && this.OverrideTileIds.Count > 0
        && this.OverrideTileIds.Contains tile
        && not <| this.OverrideTileIds.Contains neighbor

    let isOverrideTile (tile: int) (this: HexTileDataOverrider) =
        this.EditMode && this.OverrideTileIds.Contains tile

    let isOverrideNoRiver (tile: int) (this: HexTileDataOverrider) =
        isOverrideTile tile this && this.RiverMode = OptionalToggle.No

    let isOverrideNoRoad (tile: int) (this: HexTileDataOverrider) =
        isOverrideTile tile this && this.RoadMode = OptionalToggle.No

    let elevation (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplyElevation then
            this.ActiveElevation
        else
            TileValue.elevation tileValue

    let getEdgeType
        (tile1: int)
        (tileValue1: TileValue)
        (tile2: int)
        (tileValue2: TileValue)
        (this: HexTileDataOverrider)
        =
        HexMetrics.getEdgeType
        <| elevation tile1 tileValue1 this
        <| elevation tile2 tileValue2 this

    let waterLevel (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplyWaterLevel then
            this.ActiveWaterLevel
        else
            TileValue.waterLevel tileValue

    let isUnderwater (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this then
            waterLevel tile tileValue this > elevation tile tileValue this
        else
            TileValue.isUnderwater tileValue

    let streamBedY (tile: int) (tileValue: TileValue) (unitHeight: float32) (this: HexTileDataOverrider) =
        if isOverrideTile tile this then
            (float32 (elevation tile tileValue this) + HexMetrics.streamBedElevationOffset)
            * unitHeight
        else
            TileValue.streamBedY unitHeight tileValue

    let riverSurfaceY (tile: int) (tileValue: TileValue) (unitHeight: float32) (this: HexTileDataOverrider) =
        if isOverrideTile tile this then
            (float32 (elevation tile tileValue this) + HexMetrics.waterElevationOffset)
            * unitHeight
        else
            TileValue.streamBedY unitHeight tileValue

    let waterSurfaceY (tile: int) (tileValue: TileValue) (unitHeight: float32) (this: HexTileDataOverrider) =
        if isOverrideTile tile this then
            (float32 (waterLevel tile tileValue this) + HexMetrics.waterElevationOffset)
            * unitHeight
        else
            TileValue.waterSurfaceY unitHeight tileValue

    let hasRivers (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this && TileFlag.hasRivers tileFlag

    let hasIncomingRiver (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this && TileFlag.hasIncomingRiver tileFlag

    let hasOutgoingRiver (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this && TileFlag.hasOutgoingRiver tileFlag

    let hasRiverBeginOrEnd (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this && TileFlag.hasRiverBeginOrEnd tileFlag

    let hasRiverThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this && TileFlag.hasRiverThroughEdge dir tileFlag

    let hasIncomingRiverThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) (this: HexTileDataOverrider) =
        not <| isOverrideNoRiver tile this
        && TileFlag.hasIncomingRiverThoughEdge dir tileFlag

    let hasRoads (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        not <| isOverrideNoRoad tile this && TileFlag.hasRoads tileFlag

    let hasRoadThroughEdge (tile: int) (tileFlag: TileFlag) (dir: int) (this: HexTileDataOverrider) =
        not <| isOverrideNoRoad tile this && TileFlag.hasRoadThroughEdge dir tileFlag

    let walled (tile: int) (tileFlag: TileFlag) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.WalledMode <> OptionalToggle.Ignore then
            this.WalledMode = OptionalToggle.Yes
        else
            TileFlag.walled tileFlag

    let urbanLevel (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplyUrbanLevel then
            this.ActiveUrbanLevel
        else
            TileValue.urbanLevel tileValue

    let farmLevel (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplyFarmLevel then
            this.ActiveFarmLevel
        else
            TileValue.farmLevel tileValue

    let plantLevel (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplyPlantLevel then
            this.ActivePlantLevel
        else
            TileValue.plantLevel tileValue

    let specialIndex (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        if isOverrideTile tile this && this.ApplySpecialIndex then
            this.ActiveSpecialIndex
        else
            TileValue.specialIndex tileValue

    let isSpecial (tile: int) (tileValue: TileValue) (this: HexTileDataOverrider) =
        specialIndex tile tileValue this <> 0
